bpf/process/types/basic.h

```diff
@@ -2161,25 +2161,28 @@ do_actions(void *ctx, struct msg_generic_kprobe *e, struct selector_action *acti
 }
 
 static inline __attribute__((always_inline)) long
-filter_read_arg(void *ctx, int index, struct bpf_map_def *heap,
+filter_read_arg(void *ctx, struct bpf_map_def *heap,
 		struct bpf_map_def *filter, struct bpf_map_def *tailcalls,
 		struct bpf_map_def *config_map)
 {
 	struct msg_generic_kprobe *e;
 	struct event_config *config;
-	int pass, zero = 0;
+	int index, pass, zero = 0;
 
 	e = map_lookup_elem(heap, &zero);
 	if (!e)
 		return 0;
 	config = map_lookup_elem(config_map, &e->idx);
 	if (!config)
 		return 0;
-	pass = filter_args(e, index, filter, config->flags & FLAGS_EARLY_FILTER);
+	index = e->index;
+	pass = filter_args(e, index & 7, filter, config->flags & FLAGS_EARLY_FILTER);
 	if (!pass) {
 		index++;
-		if (index <= MAX_SELECTORS && e->sel.active[index])
-			tail_call(ctx, tailcalls, MIN_FILTER_TAILCALL + index);
+		if (index <= MAX_SELECTORS && e->sel.active[index & 7]) {
+			e->index = index;
+			tail_call(ctx, tailcalls, MIN_FILTER_TAILCALL);
```

Comment:
Another thing it would be readable for new comers if the programs are identified by constant during the tail_call(), or an enum with a max as _MAX_TG_TAIL_CALL_INDEX. Please note no strong opinion, just to make tail_call readable ;-)

Reply:
good idea, enum values should be better I think